aptos-move/aptos-vm/src/block_executor/mod.rs

```diff
@@ -0,0 +1,190 @@
+
+    fn get_deltas(&self) -> Vec<(StateKey, DeltaOp)> {
+        self.0
+            .delta_change_set()
+            .iter()
+            .map(|(key, op)| (key.clone(), *op))
+            .collect()
+    }
+
+    /// Execution output for transactions that comes after SkipRest signal.
+    fn skip_output() -> Self {
+        Self(TransactionOutputExt::from(TransactionOutput::new(
+            WriteSet::default(),
+            vec![],
+            0,
+            TransactionStatus::Retry,
+        )))
+    }
+}
+
+pub struct BlockAptosVM();
+
+impl BlockAptosVM {
+    fn process_parallel_block_output<S: StateView>(
+        results: Vec<AptosTransactionOutput>,
+        delta_resolver: OutputDeltaResolver<StateKey, WriteOp>,
+        state_view: &S,
+    ) -> Vec<TransactionOutput> {
+        // TODO: MVHashmap, and then delta resolver should track aggregator keys.
+        let mut aggregator_keys: HashMap<StateKey, anyhow::Result<ResolvedData>> = HashMap::new();
+        for res in results.iter() {
+            let output_ext = AptosTransactionOutput::as_ref(res);
+            for (key, _) in output_ext.delta_change_set().iter() {
+                if !aggregator_keys.contains_key(key) {
+                    aggregator_keys.insert(key.clone(), state_view.get_state_value(key));
+                }
+            }
+        }
+
+        let materialized_deltas =
+            delta_resolver.resolve(aggregator_keys.into_iter().collect(), results.len());
+
+        results
+            .into_iter()
+            .zip(materialized_deltas.into_iter())
+            .map(|(res, delta_writes)| {
+                let output_ext = AptosTransactionOutput::into(res);
+                output_ext.output_with_delta_writes(WriteSetMut::new(delta_writes))
+            })
+            .collect()
+    }
+
+    fn process_sequential_block_output(
+        results: Vec<AptosTransactionOutput>,
+    ) -> Vec<TransactionOutput> {
+        results
+            .into_iter()
+            .map(|res| {
+                let output_ext = AptosTransactionOutput::into(res);
+                let (deltas, output) = output_ext.into();
+                debug_assert!(deltas.is_empty(), "[Execution] Deltas must be materialized");
+                output
+            })
+            .collect()
+    }
+
+    pub fn execute_block<S: StateView>(
+        transactions: Vec<Transaction>,
+        state_view: &S,
+        concurrency_level: usize,
+    ) -> Result<Vec<TransactionOutput>, VMStatus> {
+        // Verify the signatures of all the transactions in parallel.
+        // This is time consuming so don't wait and do the checking
+        // sequentially while executing the transactions.
+        let signature_verified_block: Vec<PreprocessedTransaction> =
+            RAYON_EXEC_POOL.install(|| {
+                transactions
+                    .par_iter()
+                    .map(|txn| preprocess_transaction::<AptosVM>(txn.clone()))
+                    .collect()
+            });
```

Comment:
RAYON_EXEC_POOL here is the same threadpool that we use in parallel execution. It has num_cpu threads but parallel execution uses concurrency_level many of them (by dispatching long-lived tasks).

It makes sense to use the same threadpool for other execution-related work, instead of going to the global rayon threadpool (which was happing before) as that could be less predictable and make performance unpredictable/unreliable.